code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+        
+        baz = {1: {2: {3: "hello"}}}
+    
+
+    <textarea name="form[q2.3]" rows="1" required=""></textarea>
+
+<details markdown="1"><summary  markdown="span">Type variables</summary>
+
+Generics work through the use of type variables. In Python these variables are provided by `TypeVar` from the `typing` module. Here is how it works:
+
+
+    from typing import TypeVar
+
+    T = TypeVar('T')  # Can be anything
+    N = TypeVar('N', int, float)  # Must be int or float
+
+
+Type variables can be unconstraint, like `T` above. In this case `T` can be any type at all. Or type variables can be constraint, like `N` above. In which case `N` can only be an `int` or a `float`. Type variables can come in place of actual types to create for instance generic functions:
+
+
+    from typing import Iterable, TypeVar
+
+    T = TypeVar('T')
+
+    def first(items: list[T]) -> T:
+        return items[0]
+
+
+`first` will return the first item in the list, but what type is returned is dependent on the list. For instance, if `first` is called like so:
+
+
+    n = first([1,2,3])
+
+
+Then `n` will be of type `int`. Because a `list[int]` is passed in and `T` will take on the form of `int`. `T` is what is ultimately returned from `first` and that is then why `n` is an `int`.
+
+Type variables can be used outside generic data structures too, for instance:
+
+
+    def longest(a: T, b: T) -> T:
+        return a if len(a) >= len(b) else b
+
+
+This function will work for any type T, and it will return that same type.
+
+</details>
+
+3. Annotate the generic function below with a type variable:
+
+        
+        def repeat(x, n):
+            return [x] * n
+    
+
+    <textarea name="form[q3]" rows="2" required=""></textarea>
+
+## Abstract types
+
+<details markdown="1"><summary  markdown="span">Structural subtyping aka duck typing</summary>
+
+So far we have looked at concrete types, such as integers, strings and lists. These types are expressive, you know exactly what you are working with. But, often these concrete types limit design. Take for instance this function:
+
+
+    def sum(items: list[int]) -> int:
+        total = 0
+        for item in items:
+            total += item
+        return item
+
+
+There is no reason this implementation cannot work with other types of data structures. A tuple of integers or a set of integers should work just fine, but the type hint `list[int]` will only accept a concrete `list`. This is quite unpythonic!
+
+Looking at the implementation of `sum`, all that is needed from `items` is that it works with a for-loop. Or more precisely, the data structure needs to be iterable. In this case we only care about a property of the type, not the concrete thing. Rather, if the type we insert into the function is somewhat list-like, the function should work just fine. In comes duck typing:
+
+> if it walks like a duck, swims like a duck, and quacks like a duck... it's a duck.
```

Comment:
Ik twijfel of "duck typing" hier goed tot z'n recht komt. Ik moet deze tekst nog even goed lezen, het is relatief veel informatie tegelijk ten opzichte van de info bij de andere oefeningen.

Reply:
Ik heb het nu opgesplitst in twee alinea's en iets meer lucht toegevoegd

Reply:
Ik vermoed dat je het meer to-the-point kunt overbrengen als je juist protocols/interfaces noemt in de hoofduitleg, en dan duck typing "for the interested student".

Het punt is denk ik dat bij "duck typing" een soort foute analogie zit, "het kwaakt dus het is een eend" terwijl we bij programmeren zelf een kunstmatige wereld creëren en zelf kunnen bepalen hoe dingen werken (en welk geluid ze maken).

Maar ik heb het nog steeds niet top-to-bottom gelezen, ik denk alvast hardop.

Reply:
Ik vind juist duck typing mooi hier omdat het zo makkelijk blijft hangen. 

> terwijl we bij programmeren zelf een kunstmatige wereld creëren 

Niet per se mee eens. Ja we bouwen het zelf, maar dat kan ook door een collega uit de andere kant van de wereld zijn. En ik heb echt niet zijn eenden onder controle hier 😂

Reply:
Kennen we nog iemand met wat grafische skills? 

> Duck typing is great and all, but what if we actually do want a duck, not something that happens to act like a duck.

Reply:
Ik zou alsnog de volgorde ietsje aanpassen om het scherper te maken:

- Deze functie moet met meer dan alleen een lijst werken
- Deze functie moet met elke collectie van waardes werken
- Deze functie moet met elke variabele werken waar we for-in mee kunnen gebruiken
- Intern in Python betekent dat dat de variabele de methode `__iter__` moet hebben
- We kunnen dit aangeven door het (abstracte) type Iterable, die belooft dat de functie werkt met elk object dat een `__iter__`-methode heeft.
- Dit fenomeen heet duck typing: ...

Reply:
Ik ben meer fan van de huidige volgorde, in abstracte zin:

* Probleem creeren, namelijk een ` list` als input legt onnodige ristricties op
* Wens uitspreken, eigenlijk hoeft de input alleen maar om te gaan met een for-loop
* Oplossing erbij trekken, duck typing
* Oplossing in Python laten zien
* Korte toelichting van hoe dit werkt underneath the hood, `__iter__` 

Zo hoef ik niet te onderbouwen waarom deze functie met meer dan alleen een lijst moet werken. Eerder, het kan al, dus waarom restricties opleggen.

Reply:
Ik heb deze alinea's een stuk preciezer gemaakt (hopelijk!) hieronder. Ik denk dat het inderdaad in die volgorde kan blijven, maar door wat meer expliciete verbanden te leggen maak ik het voor mezelf wel een stuk begrijpelijker, en hopelijk ook voor de studenten. Met name heb ik in de derde alinea heel expliciet en heel precies duck typing geprobeerd te definiëren (jouw ideeën hergebruikend).

---

There is no reason this implementation cannot work with other types of data structures that contain multiple values of the same type. For example, a tuple of integers and a set of integers could just as well be summed using the same algorithm. However, our type hint `list[int]` will specifically only accept a list. This is quite un-pythonic!

Looking at the implementation of `sum`, all that we need from the parameter `items` is that it allows extracting all its elements using a for-loop. In the language of Python, we say that the data structure needs to be *iterable*.

So we would like the type hint to be less strict. Instead of specifying `list` as the concrete type, we would only like to be able to specify that the method works with anything that is iterable and contains integers. This is called duck typing: we aim to specify the thing that we need the object to do, and not wat concrete type it has.

> Duck typing is derived from the duck test: _"if it walks like a duck, swims like a duck, and quacks like a duck… it’s a duck."_

Well, we don't need a duck, in our case, but an object that is iterable. Luckily Python's `typing` module comes with a bunch of "duck types" built-in, one of which is indeed `Iterable`. We can use it like so: